File: 64-minimum-path-sum/minimum-path-sum.java
class Solution {

    private int solve(int m, int n, int[][] grid, int[][] dp){
        if(m < 0 || n < 0) return Integer.MAX_VALUE;
        if(m==0 && n ==0) return grid[m][n];
        if(dp[m][n]!=-1) return dp[m][n];
        int min = Math.min(solve(m-1,n,grid, dp), solve(m,n-1,grid, dp));
        if(min == Integer.MAX_VALUE) return min;
        return dp[m][n] = grid[m][n] + min; 
    }
    public int minPathSum(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int[][] dp = new int[m+1][n+1];
        for(int i=0;i<m;i++){
            Arrays.fill(dp[i],-1);
        }
        return solve(m-1,n-1,grid, dp);
    }
}

File: 64-minimum-path-sum/minimum-path-sum.cpp
class Solution {
public:
    int solve(int m, int n, vector<vector<int>>&grid, vector<vector<int>>&dp){
        if( m < 0 || n < 0) return INT_MAX;
        if(m==0 && n ==0) return grid[0][0];
        if(dp[m][n]!=-1) return dp[m][n];
        return dp[m][n] = grid[m][n] +  min( solve(m-1,n,grid, dp), solve(m,n-1,grid, dp) );
    }
    int minPathSum(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<int>>dp(m, vector<int>(n,-1));
        return solve(m-1,n-1,grid,dp);
    }
};